oracle/src/confirmRelay.js

```diff
@@ -0,0 +1,155 @@
+  try {
+    if (connection.isConnected()) {
+      if (config.maxProcessingTime) {
+        await watchdog(() => main({ sendToQueue, sendToWorker, txHash }), config.maxProcessingTime, () => {
+          logger.fatal('Max processing time reached')
+          process.exit(EXIT_CODES.MAX_TIME_REACHED)
+        })
+      } else {
+        await main({ sendToQueue, sendToWorker, txHash })
+      }
+    } else {
+      setTimeout(() => {
+        runMain({ sendToQueue, sendToWorker })
+      }, config.pollingInterval)
+    }
+  } catch (e) {
+    logger.error(e)
+  }
+}
+
+function processEvents(events) {
+  switch (config.id) {
+    case 'native-erc-signature-request':
+    case 'erc-erc-signature-request':
+    case 'erc-native-signature-request':
+      return processSignatureRequests(events)
+    case 'native-erc-collected-signatures':
+    case 'erc-erc-collected-signatures':
+    case 'erc-native-collected-signatures':
+      return processCollectedSignatures(events)
+    case 'native-erc-affirmation-request':
+    case 'erc677-erc677-affirmation-request':
+    case 'erc-native-affirmation-request':
+    case 'erc-erc-affirmation-request':
+      return processAffirmationRequests(events)
+    case 'erc-erc-transfer':
+    case 'erc-native-transfer':
+      return processTransfers(events)
+    case 'amb-signature-request':
+      return processAMBSignatureRequests(events)
+    case 'amb-collected-signatures':
+      return processAMBCollectedSignatures(events)
+    case 'amb-affirmation-request':
+      return processAMBAffirmationRequests(events)
+    default:
+      return []
+  }
+}
+
+async function checkConditions() {
+  let state
+  switch (config.id) {
+    case 'erc-native-transfer':
+      logger.debug('Getting token address to listen Transfer events')
+      state = await getTokensState(bridgeContract, logger)
+      updateEventContract(state.bridgeableTokenAddress)
+      break
+    case 'erc-native-half-duplex-transfer':
+      logger.debug('Getting Half Duplex token address to listen Transfer events')
+      state = await getTokensState(bridgeContract, logger)
+      updateEventContract(state.halfDuplexTokenAddress)
+      break
+    default:
+  }
+}
+
+function updateEventContract(address) {
+  if (eventContractAddress !== address) {
+    eventContractAddress = address
+    eventContract = new web3Instance.eth.Contract(config.eventAbi, eventContractAddress)
+  }
+}
+
+async function main({ sendToQueue, txHash }) {
+  try {
+    await checkConditions()
+
+    const events = await getEventsFromTx({
+      web3: web3Instance,
+      contract: eventContract,
+      event: config.event,
+      txHash,
+      filter: config.eventFilter
+    })
+    logger.info(`Found ${events.length} ${config.event} events`)
+
+    if (events.length) {
+      const job = await processEvents(events)
+      logger.info('Transactions to send:', job.length)
+
+      if (job.length) {
+        await sendToQueue(job)
```

Comment:
Does it mean that we need to have the oracle up and running? Could you estimate how much it could take to have two options: the oracle has been run before `confirmRelay.js` is invoked and the oracle has been stopped. It will be useful in the future to have a script which could operate with the signer private key without need to deploy and run entire oracle.

Reply:
I have extended the script in the following way:
* if `ORACLE_QUEUE_URL` is available, the script will use running oracle's environment to send a tx (it will add a new job in a queue).
* if `ORACLE_QUEUE_URL` is not available, the script will send a tx by itself using provided private key.

For the second option, the script can be simply run by `docker run --env-file oracle/.env poanetwork/tokenbridge-oracle:latest yarn confirm:transfer 0x76..a5`